File: algorithms/[52]N 皇后 II/solution.py
# topics = ["回溯算法"]

from typing import List


class Solution:
    def totalNQueens(self, n: int) -> int:
        res = 0
        path: List[int] = []

        def backtrack(f: int = 0) -> None:
            nonlocal res
            if f == n:
                res += 1
                return
            for i in range(n):
                if is_valid(i):
                    path.append(i)
                    backtrack(f + 1)
                    path.pop()

        # 判断是否互不攻击
        def is_valid(i: int) -> bool:
            if not path:
                return True
            for idx, col in enumerate(path):
                if i == col or abs(i - col) == abs(len(path) - idx):
                    return False
            return True

        backtrack()

        return res

File: algorithms/[52]N 皇后 II/solution.ts
// topics = ["回溯算法"]

function totalNQueens(n: number): number {
  let res = 0;
  const path: number[] = [];

  const backtrack = (f = 0): void => {
    if (f == n) {
      res += 1;
      return;
    }

    for (let i = 0; i < n; i += 1) {
      if (isValid(i)) {
        path.push(i);
        backtrack(f + 1);
        path.pop();
      }
    }
  };

  const isValid = (i: number): boolean => {
    if (path.length == 0) {
      return true;
    }
    for (const [idx, col] of path.entries()) {
      if (i == col || Math.abs(i - col) == Math.abs(path.length - idx)) {
        return false;
      }
    }
    return true;
  };

  backtrack();

  return res;
}
